index.bs

```diff
@@ -3134,6 +3218,69 @@ and an optional [=boolean=] |isNoised| (default false):
     with |dataTypes|, |source|, and |isNoised|.
 
+To <dfn>find sources with common destinations and reporting origin</dfn> given an [=attribution source=] |pendingSource|:
+1. Let |matchingSources| be a new [=list=].
+1. [=set/iterate|For each=] |source| of the user agent's [=attribution source cache=]:
+    1. Let |commonDestinations| be the [=set/intersection=] of |source|'s [=attribution source/attribution destinations=] and |pendingSource|'s [=attribution source/attribution destinations=].
+    1. If |commonDestinations| [=set/is empty=], [=iteration/continue=].
+    1. If |source|'s [=attribution source/reporting origin=] and |pendingSource|'s [=attribution source/reporting origin=] are not [=same origin=], [=iteration/continue=].
+    1. [=list/Append=] |source| to |matchingSources|.
+1. Return |matchingSources|.
+
+To <dfn>check if source has incompatible attribution scope fields</dfn> given an [=attribution source=] |source| and an [=attribution source=] |pendingSource|:
+1. If |pendingSource|'s [=attribution source/attribution scope limit=] is null:
+    1. [=Assert=]: |source|' [=attribution source/attribution scope limit=] is null.
+    1. Return false.
+1. If |source|'s [=attribution source/attribution scope limit=] is null, return true.
+1. If |source|'s [=attribution source/max event states=] is not equal to |pendingSource|'s [=attribution source/max event states=], return true.
+1. If |source|'s [=attribution source/attribution scope limit=] is less than |pendingSource|'s [=attribution source/attribution scope limit=], return true.
+1. Return false.
+
+To <dfn>remove associated event reports</dfn> given an [=attribution source/source identifier=] |sourceId| and a [=moment=] |minTriggerTime|:
+1. [=set/iterate|For each=] [=event-level report=] |report| of the [=event-level report cache=]:
+    1. If |report|'s [=event-level report/source identifier=] is not equal to |sourceId|, [=iteration/continue=].
+    1. If |report|'s [=event-level report/trigger time=] is less than |minTriggerTime|, [=iteration/continue=].
+    1. [=set/Remove=] |report| from the [=event-level report cache=].
+    1. [=list/Remove=] all elements from [=attribution rate-limit records=] whose [=attribution rate-limit record/entity ID=] is equal to |report|'s [=event-level report/report ID=] and [=attribution rate-limit record/scope=] is equal to "<code>[=rate-limit scope/event-attribution=]</code>".
+
+To <dfn>remove sources with attribution scopes not selected for destination</dfn> given a [=site=] |destination| and an [=attribution source=] |pendingSource|:
+1. Let |attributionScopeRecords| be a new [=list=].
+1. [=set/iterate|For each=] |attributionScope| in |pendingSource|'s [=attribution source/attribution scopes=]:
+    1. [=list/Append=] the [=tuple=] (|attributionScope|, |pendingSource|'s [=attribution source/source time=], |pendingSource|'s [=attribution source/source identifier=]) to |attributionScopeRecords|.
```

Comment:
In general, the scopes of the pending source should always be selected as they have most recent time. But do we need to worry about the case with bad time in the cache?

Reply:
cc @apasel422 for another time-related issue.

Reply:
cc @ThomasQuesadilla

Reply:
Following up on this. With the current algorithm, it's possible that the pending source may be deleted  in the case of user misconfigured clocks. If we want to ensure that the scopes of the pending source are always selected, we may consider inserting them into the set directly and skip the sorting step.

Reply:
Why do we want to assume something specific like this? Shouldn't *all* assertions about time be based on a consistent reference? If the user is amending their clock, isn't it their prerogative?

Reply:
Alternatively,  if we don't make this assumption, then we need to handle the case that the pending source is deleted. Either is fine with me.